### PR TITLE
Use filter strategy for streaming async server

### PR DESCRIPTION
**Pull request by bondolo**
Motivation:
Currently the filter strategy is not considered in the computation of
the effective strategy for async streaming servers. This was done
because with a default builder strategy the result could be less
offloading which would be unexpected for that builder strategy.
Modifications:
Consider filter strategy in computing effective strategy for
async streaming servers.
Result:
More predictable offloading behavior.

### CODE REVIEW

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java [hunk 1]
```diff
@@ -356,11 +347,13 @@ private HttpExecutionContext buildExecutionContext(final HttpExecutionStrategy s
...
 
-    private HttpExecutionStrategy computeServiceStrategy(Object service) {
-        HttpExecutionStrategy serviceStrategy = requiredOffloads(service, defaultStrategy());
+    private HttpExecutionStrategy computeServiceStrategy(HttpExecutionStrategyInfluencer service) {
+        HttpExecutionStrategy serviceStrategy = service.requiredOffloads();
         HttpExecutionStrategy filterStrategy = computeRequiredStrategy(serviceFilters, serviceStrategy);
-        return defaultStrategy() == strategy ? filterStrategy :
-                strategy.hasOffloads() ? strategy.merge(filterStrategy) : offloadNone();
+        return defaultStrategy() == strategy ?
+            filterStrategy.hasOffloads() ? strategy.merge(filterStrategy) : strategy :
```
Comment (bondolo) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java:354]: I am tempted to replace the return of `strategy` when `filterStrategy` has no offloads with `offloadAll()`. Currently this results in `defaultStrategy` appearing the context which will merge differently with non-default execution strategy than `offloadAll()` and potentially result in additional offloading.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java [hunk 2]
```diff
@@ -261,7 +253,8 @@ public HttpServerBuilder bufferAllocator(final BufferAllocator allocator) {
 
     @Override
     public Single<HttpServerContext> listenStreaming(final StreamingHttpService service) {
```
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java:255]: Consider adding a test when users specify a custom strategy with one flag on the builder and another flag on the filter. Assert that the resulting strategy has both offloading flags set for any `listen*` method.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java [hunk 3]
```diff
@@ -356,11 +349,13 @@ private HttpExecutionContext buildExecutionContext(final HttpExecutionStrategy s
...
 
-    private HttpExecutionStrategy computeServiceStrategy(Object service) {
-        HttpExecutionStrategy serviceStrategy = requiredOffloads(service, defaultStrategy());
+    private HttpExecutionStrategy computeServiceStrategy(HttpExecutionStrategyInfluencer service) {
+        HttpExecutionStrategy serviceStrategy = service.requiredOffloads();
         HttpExecutionStrategy filterStrategy = computeRequiredStrategy(serviceFilters, serviceStrategy);
-        return defaultStrategy() == strategy ? filterStrategy :
-                strategy.hasOffloads() ? strategy.merge(filterStrategy) : offloadNone();
+        return defaultStrategy() == strategy ?
+            filterStrategy.hasOffloads() ? strategy.merge(filterStrategy) : strategy :
```
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java:356]: `strategy.merge(filterStrategy)` - this will result in a warning log printed by `DEFAULT_HTTP_EXECUTION_STRATEGY`. Should it be `filterStrategy.hasOffloads() ? filterStrategy : strategy`?

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java [hunk 4]
```diff
@@ -356,11 +349,13 @@ private HttpExecutionContext buildExecutionContext(final HttpExecutionStrategy s
...
+            // We currently treat the builder strategy specially, allowing it to override the filter strategy.
+            strategy.hasOffloads() ? strategy : offloadNone();
```
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java:358]: In this case, looks like by some reasons users passed a custom strategy to the builder that does not have offloading. As a user, I would expect to see my custom strategy on the `ServerContext`. `offloadNone()` will be surprising to me. Maybe I have a special subtype that carries more information/state.

Comment (bondolo) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java:358]: The strategy that is exposed is in the `ExecutionContext` currently almost always `OFFLOAD_EVENT_CLOSE_STRATEGY`  because there is always a non-offloading filter (`ClearAsyncContext`) and the real offloading is done by the `OffloadingFilter`.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java (hunk 3 above)
Comment (bondolo) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java:356]: yes, well spotted. this can be just `filterStrategy`.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java (hunk 4 above)
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java]: I mean that users can implement their own `HttpExecutionStrategy`, keep `hasOffloads()` as `false`, but have some other flags or state. Don't think replacing it with `offloadNone()` is reasonable. I would expect this to be:
```java
return defaultStrategy() == strategy ?
            (filterStrategy.hasOffloads() ? filterStrategy : strategy) :
            (filterStrategy.hasOffloads() ? strategy.merge(filterStrategy) : strategy);
```

Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java]: Same but in different way:
```java
return filterStrategy.hasOffloads() ?
            (defaultStrategy() == strategy ? filterStrategy : strategy.merge(filterStrategy)) : strategy;
```

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/BlockingStreamingToStreamingService.java [hunk 5]
```diff
@@ -57,6 +57,11 @@
         this.original = requireNonNull(original);
     }
 
+    @Override
+    public HttpExecutionStrategy requiredOffloads() {
+        return original.requiredOffloads();
```
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/BlockingStreamingToStreamingService.java:62]: Hm... might be wrong, but it looks a bit inconsistent. Check out the ctor. It can optimize the strategy and pass the result to super class to be used for `serviceInvocationStrategy()`. The value for `requiredOffloads()` will be different. We may need to reconsider the whole approach with `serviceInvocationStrategy()` later.

Is this change required for this PR? If not, let's revert and work on it separately.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java [hunk 6]
```diff
@@ -289,12 +283,16 @@ private HttpExecutionContext buildExecutionContext(final HttpExecutionStrategy s
...
+     * @param apiStrategy The {@link HttpExecutionStrategy} provided by the user/API
      * @return A {@link Single} that completes when the server is successfully started or terminates with an error if
      * the server could not be started.
      */
     private Single<HttpServerContext> listenForService(final StreamingHttpService rawService,
-                                                       final HttpExecutionStrategy serviceStrategy) {
+                                                       final HttpExecutionStrategy apiStrategy) {
+        HttpExecutionStrategy serviceStrategy = computeServiceStrategy(rawService);
+        HttpExecutionStrategy serverStrategy =
+                apiStrategy.hasOffloads() ? apiStrategy.merge(serviceStrategy) : apiStrategy;
```
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java:287]: 1. Consider adding an assertion that `apiStrategy` is never `defaultStrategy()`
2. When `apiStrategy.hasOffloads() == false`, why do we ignore the result of `computeServiceStrategy`?

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/BlockingStreamingToStreamingService.java (hunk 5 above)
Comment (bondolo) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/BlockingStreamingToStreamingService.java:62]: This change is a simplification to better handle the builder `defaultStrategy()` with the API conversions and each having their own default. The builder strategy is now passed to the API conversion and the default, if appropriate, is applied in the conversion. The conversion then returns either the appropriate default strategy or the builder strategy. The builder strategy overrides the default.

This is then later combined with the filter and service strategy in the `listenForService` which allows the service and filters to adjust the strategy and add additional offloading if needed.

Comment (idelpivnitskiy) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/BlockingStreamingToStreamingService.java:62]: Ok, can you please add some clarification comments in code to describe why `requiredOffloads()` is different from `serviceInvocationStrategy()`? 
We introduced `serviceInvocationStrategy()` in the past because `influenceStrategy` required an input strategy. With `requiredOffloads()` there is a change to remove `serviceInvocationStrategy()` in future.

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java [hunk 7]
```diff
@@ -307,9 +307,28 @@ HttpServerBuilder appendServiceFilter(Predicate<StreamingHttpRequest> predicate,
...
+     *  builder:                service:                  filters:             Effective:
+     *  defaultStrategy()
+     *  (if service or filter needs additional offloading, add it)
+     *                          offloadNone()             offloadNone()        apiDefaultStrategy()
+     *                          customStrategy()          offloadNone()        apiDefaultStrategy() + customStrategy()
+     *                          offloadNone()             customStrategy()     apiDefaultStrategy() + customStrategy()
+     *  customStrategy()
+     *  (if service or filter needs additional offloading, add it)
+     *                          offloadNone()             offloadNone()        customStrategy()
+     *                          customStrategy()          offloadNone()        customStrategy() + customStrategy()
```
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java:324]: This change looks unrelated to the original problem this PR is trying to solve and increases review iterations. Consider taking it out to a separate PR.
These abbreviations are still hard to understand for new users.

Comment (bondolo) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java]: Sorry, I should have moved the PR in to draft before pushing those changes. I have replaced the PR with a different more minimal one.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java (hunk 6 above)
Comment (bondolo) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java]: The behaviour of the builder `executionStrategy()` is effectively to override the filter and service strategy. In this case the builder strategy was likely either `offloadNone()` or `offloadNever()` indicating that the user desired no offloading.

File: servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/BasicAuthStrategyInfluencerTest.java [hunk 8]
```diff
@@ -146,8 +146,7 @@ private BlockingHttpClient setup(boolean noOffloadsInfluence) throws Exception {
         return this.client;
     }
 
-    private static final class OffloadCheckingService implements StreamingHttpService,
-                                                                 ExecutionStrategyInfluencer<HttpExecutionStrategy> {
```
Comment (bondolo) [inline servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/BasicAuthStrategyInfluencerTest.java:150]: `ExecutionStrategyInfluencer<HttpExecutionStrategy>` is now redundant.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java (hunk 4 above)
Comment (bondolo) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java]: no longer applicable

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java [hunk 9]
```diff
@@ -261,7 +262,11 @@ public HttpServerBuilder bufferAllocator(final BufferAllocator allocator) {
 
     @Override
     public Single<HttpServerContext> listenStreaming(final StreamingHttpService service) {
-        return listenForService(service, strategy);
+        HttpExecutionStrategy serviceStrategy = requiredOffloads(service, defaultStrategy());
+        HttpExecutionStrategy filterStrategy = computeRequiredStrategy(serviceFilters, serviceStrategy);
+        HttpExecutionStrategy useStrategy = defaultStrategy() == strategy ? offloadAll() :
+                strategy.hasOffloads() ? strategy.merge(filterStrategy) : offloadNone();
```
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java:268]: Why does this have to be different from all other API variants? I mean, why don't we use `computeServiceStrategy` method here? Let's discuss and we should at least leave a comment in the code to clarify the difference.

My main concerning part is `defaultStrategy() == strategy ? offloadAll()`. If I'm a user who has 0 filters and passed `StreamingHttpService` impl that override `requireOffloads()` to `offloadNone()`, I won't expect `offloadAll()` by default.

Comment (bondolo) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultHttpServerBuilder.java:268]: We can't use `computeServiceStrategy()` because it maps `defaultStrategy()` which we want to handle differently. The other three variants with adapters have their only handling for `defaultStrategy()` and won't ever trigger the `defaultStrategy() == strategy` in `computeServiceStrategy()`

> My main concerning part is `defaultStrategy() == strategy ? offloadAll()`. If I'm a user who has 0 filters and passed `StreamingHttpService` impl that override `requiredOffloads()` to `offloadNone()`, I won't expect `offloadAll()` by default.

Fair point. If the service does not override `requiredOffloads()` then the default `offloadAll()` it provides is sufficient to ensure safety otherwise builder `defaultStrategy()` allows whatever strategy the filter and service require.

With this change, and the additional change of returning `strategy` when `strategy.hasNoOffloads()` is true the result is that we can use `computeServiceStrategy()` everywhere.

**Review by idelpivnitskiy — APPROVED** [review]
Thank you!
